### DISCUSSION TITLE
Swerve modules turning at full speed whenever any input is pressed

### DISCUSSION THREAD

**Discussion by wisplite** [BroncBotz3481/YAGSL-Example#114] [Q&A] [answered]
Hello! I'm the lead programmer for Team 8234, and we've been trying to get our swerve working with this library with no luck. We're using all NEO motors with attached throughbore encoders for the angle motors, and the Pigeon 2 IMU. Currently whenever an input is pressed, the angle motors rotate at full speed, despite the desired direction staying still in the swerve display of frc-web-components.

Here is a video demonstrating the behavior: https://youtube.com/watch?v=H-WVocYqxzI

Any help with this issue would be greatly appreciated!

**Comment by thenetworkgrinch** [comment]
Sorry for the late response! Thank you for the excellent video it illustrated your problem perfectly! It seems like you could need to invert all of your angle motors and/or use the correct conversion factor, remember that if NEOs are used the encoder pulse per revolution is 1!

**Comment by wisplite** [comment]
Apologies for not getting back about this sooner, our team ended up disassembling the test swerve chassis and building our final one, and we've been waiting on the belly pan. It turns out there were some wiring issues with the last chassis along with some code issues, and long story short we resolved the wheels spinning in place like that.

Despite resolving that issue, we're currently having issues with the swerve drivetrain not moving consistently, it seems to move in random directions if I hold the joystick forward, and the wheels never seem to move to the correct location.

Here are some videos of the issue and the repository for the test code:
https://github.com/tbukfrc/2024SwerveDriveTest
https://youtube.com/watch?v=CXuSM-tJMEc
https://youtube.com/watch?v=mY8j-Fb4328

If you're able to help with this, I would greatly appreciate it, it's our first year using swerve and a lot of people on our team are worried that it won't work.

**Reply by thenetworkgrinch** [reply]
That looks like you could have the wrong conversion factors or an untuned drive PID.

The desired state its sending is 0, which means there must be something very wrong with how you built it or how its reacting.

**Reply by wisplite** [reply]
I'll look into it. I'm setting up the conversion factor in Java right now, but I'll set it up in the JSON and make sure it looks right. I should have ~3 hours to work on this tomorrow, so hopefully I can figure it out then.

**Reply by wisplite** [reply]
I ended up staying late at our school and even after recalculating everything we're having the same issues. I'm not really sure what it could be, I've looked over the code multiple times and I can't find anything that is obviously wrong.

**Reply by thenetworkgrinch** [reply]
What modules are you using? 
Have you tested your drive motors independently?
Have you tested your angle motors independently?
Are you sure that you put in the right types and ID's?

**Reply by thenetworkgrinch** [reply]
I just took a look at your configuration and noticed your absolute encoder offsets are 0. You say that you are using throughbore's not canandcoders. You need to set the `absoluteEncoderOffset` in the JSON files instead of the REV hardware client. 

As for your conversion factors, you should use the standard ones if you are using a standard swerve module.
https://yagsl.gitbook.io/yagsl/configuring-yagsl/standard-conversion-factors

This should also not be this low, set a usable amount please.
https://github.com/tbukfrc/2024SwerveDriveTest/blob/main/src/main/java/frc/robot/subsystems/SwerveSubsystem.java#L25

You are also using an outdated version of YAGSL, please update to `2024.4.7`
https://github.com/tbukfrc/2024SwerveDriveTest/blob/main/vendordeps/yagsl.json#L4

The changelog is here.
https://yagsl.gitbook.io/yagsl/overview/changelog

**Reply by wisplite** [reply]
After a bit of work we've got it almost working, but now the wheels all move really slowly and sometimes stop moving, and they oppose each other when rotating (every other wheel drives in the opposite direction). At one point only one of the wheels would move slowly and it would change randomly, but it got worse as we've been testing and trying to fix the rotation issue.

I've checked and it's not battery issues, the SparkMAX motor controllers are receiving a really low target output from YAGSL for some reason. One of the wheels also always has the gear facing the wrong direction, even if I add 180 to the absolute offset.

I apologize for asking so many questions, but hopefully this is the last step before we have a working swerve.

**Comment by wisplite** [comment]
After more testing, we managed to get it to move, but only robot oriented worked, field oriented would change the forward direction if we rotated. After fixing a lot of issues, however, we determined that setting absolute offsets caused all the drive motors to stop moving. If the absolute offsets are zero, the drive motors work perfectly, but the angle motors don't turn to the correct spots and the drive motors fight against each other. Do you know what could be causing this behavior? I've asked other teams and they don't seem to know, but none of them have used this library.

**Reply by thenetworkgrinch** [reply]
This issue is not unique to this library, it happens when modules are configured incorrectly. Follow this guide and it should work somewhere after step 4.

https://yagsl.gitbook.io/yagsl/configuring-yagsl/the-eight-steps

**Reply by wisplite** [reply, marked as answer]
Thank you so much! I followed your advice and our swerve chassis is working now. Our team greatly appreciates your help!
